### PR TITLE
Show matching icons based on the filename in the repo tree view

### LINKED ISSUE #11149: [Feature] Show matching icons based on the file type
Hi! It would make it much easier to be able to visually identify and distinguish files according to their file type. The current icon looks a little dull (and boring) imo.

This should cover various file types. e.g. images should have an image icon and so on..

This is how it currently looks like:
This is how it currently looks like
Here there should be a little icon for .go and a shell scrip icon for the .sh file respectively.

### PR DESCRIPTION
**Pull request by yardenshoham**
- Adds functionality that renders file and folder icons according to their name in the repo tree view
- Based on https://github.com/Claudiohbsantos/github-material-icons-extension (which is based on https://github.com/PKief/vscode-material-icon-theme)
- `build/generate-svg.js` now generate from https://github.com/PKief/vscode-material-icon-theme the:
  - icon map (`assets/material-icons.json`)
  - icon SVGs
- `EntryIcon` returns hundreds of different icons, not the old (less than 10) results
- This feature is disabled by default and can be turned on. `ui.FILE_ICONS`: **file-specific**: \[file-specific, basic\]: Whether the file icons in the repo tree view should be specific to the file name (`file-specific`) or the same icon for all files and another for the folders (`basic`)
- Closes #11149

## Before
image
image

## After (if you set `ui.FILE_ICONS = file-specific` in `app.ini`)
image
image

### CODE REVIEW

**Comment by yardenshoham** [comment]
The diff is actually pretty small if you don't count the generated files `public/img/*.svg` and `asset/material-icons.json`

File: templates/repo/view_list.tmpl
Comment (lunny) [inline templates/repo/view_list.tmpl:53]: So we don't know if it's a submodule now?

File: modules/icon/icon.go
Comment (lunny) [inline modules/icon/icon.go:48]: Maybe not everyone like the colorful icons, could we keep an app.ini configuration to switch?

**Comment by lafriks** [comment]
Test failure is related

**Comment by delvh** [comment]
Getting this PR merged will be fun as the two opposing views clash with each other in this PR:
The ones who want this change, and the ones who want to keep Gitea minimal.
I'm somewhere in between.
For me, the usefulness probably outweighs the extra resource usage.

File: modules/icon/icon.go
Comment (delvh) [reply modules/icon/icon.go:48]: However, that switch should be expanded to the file loading above.
Otherwise, you load all these extra resources, and don't use them…

File: templates/repo/view_list.tmpl
Comment (delvh) [reply templates/repo/view_list.tmpl:53]: I just had a look into the `EntryIcon` method.
That's not needed anymore as it is already checked there.
Although I do have to say that I'm only a marginal fan of the new submodule icon.

File: modules/icon/icon.go
Comment (delvh) [inline modules/icon/icon.go]: This type seems to be unused:
You deserialize it, and then you do nothing with it…

File: modules/icon/icon.go
Comment (delvh) [inline modules/icon/icon.go]:
```suggestion

	if !entry.IsDir() {
		if iconMap.FileNames[fileName] != "" {
			return iconMap.FileNames[fileName]
		}
		
		lowerFileName := strings.ToLower(fileName)
		if iconMap.FileNames[lowerFileName] != "" {
			return iconMap.FileNames[lowerFileName]
		}
		
		fileExtension := strings.TrimPrefix(path.Ext(fileName), ".")
		if iconMap.FileExtensions[fileExtension] != "" {
			return iconMap.FileExtensions[fileExtension]
		}
		
		if iconMap.LanguageIds[fileExtension] != "" {
			return iconMap.LanguageIds[fileExtension]
		}
		return "file"
	}

	if iconMap.FolderNames[fileName] != "" {
		return iconMap.FolderNames[fileName]
	}
	
	lowerFileName := strings.ToLower(fileName)
	if iconMap.FolderNames[lowerFileName] != "" {
		return iconMap.FolderNames[lowerFileName]
	}
```

File: modules/icon/icon.go
Comment (delvh) [reply modules/icon/icon.go]: Not even speaking about these other things I see in the json file that seem to be missing completely, i.e. high contrast.

File: modules/icon/icon.go
Comment (delvh) [reply modules/icon/icon.go:48]: What about `[ui].FILE_ICONS = file-specific|basic` as key with `file-specific` as default value?

File: templates/repo/view_list.tmpl
Comment (yardenshoham) [reply templates/repo/view_list.tmpl:53]: The entire icon logic (including this check) will be handled in the `EntryIcon` function

File: build/generate-svg.js
Comment (delvh) [inline build/generate-svg.js]: Seems a bit like overkill to me to clone a whole repo, copy all files, and then delete it again.
Is there no better way to achieve that?

File: modules/icon/icon.go
Comment (yardenshoham) [reply modules/icon/icon.go:48]: ~If it's not too hard I will try to include it in the user's appearance settings. Do you (or anyone else) have a high-level instruction list on how to add a setting to the user's appearance settings page? (settings? forms? templates? migrations?)~

I'll do it in `app.ini`, so instance-wide. User settings can be a PR of its own

**Review by delvh — COMMENTED** [review]
Before you do anything else, @lunny and I meant this a bit different.
Here's my temporary review.

File: build/generate-svg.js
Comment (yardenshoham) [reply build/generate-svg.js]: The key thing here is the `npm run build` that generates the JSON. It's not published so we have to generate it. I don't mind maintaining a git repo with a cron action that generates these files to some object storage if you prefer... Then here we will download a tar with the JSON + icons

File: modules/icon/icon.go
Comment (yardenshoham) [reply modules/icon/icon.go]: Thanks, I'll drop what I'm not using. Yes, the file has more features including icons for light themes. In this PR I will focus only on adding the default set of icons, a following PR could make use of the full capabilities of this icon set.

File: build/generate-svg.js
Comment (yardenshoham) [reply build/generate-svg.js]: So stick with this or the other approach with the new repo?

**Comment by silverwind** [comment]
material theme seems a bit opinionated to me. Is there a "official" vscode icon pack that we could use instead? Just trying to be as neutral as possible. Also check what GitLab is using.

**Comment by yardenshoham** [comment]
We could call this theme `material` and later add `vscode`

File: build/generate-svg.js
Comment (silverwind) [reply build/generate-svg.js]: Would prefer a icon theme that is published somewhere, ideally npm. Git repos as source are volatile and can be deleted anytime.

**Comment by silverwind** [comment]
Hmm, I think one icon theme might be enough initially. Not sure I like folder icons, it seems a bit random which folder gets a special icon and which one doesn't. [gitlab](https://gitlab.com/gitlab-org/gitlab/-/tree/master) also does not do folder icons, fwiw.

**Comment by yardenshoham** [comment]
I actually think it looks good, adds a nice touch

**Comment by yardenshoham** [comment]
This can be disabled so if someone doesn't like it it's not forced on them

**Comment by silverwind** [comment]
Checking GitLab, it does appear they use [the same icon source](https://gitlab.com/gitlab-org/gitlab-svgs/-/blob/main/build_scripts/build-svgs.sh).

**Comment by silverwind** [comment]
As per https://github.com/PKief/vscode-material-icon-theme/issues/1299#issuecomment-1061244327, there is a npm package, please use that as the source.

**Comment by silverwind** [comment]
> Sounds good, I will. Looks like we still have to run `npm run build` to generate the `material-icons.json` but it's an improvement for sure.

I'd expect the package to include that json file. Should ask the maintainer if they can add it to the npm publish.

**Comment by yardenshoham** [comment]
> I'd expect the package to include that json file. Should ask the maintainer if they can add it to the npm publish.

https://github.com/PKief/vscode-material-icon-theme/issues/1945

**Comment by yardenshoham** [comment]
I changed it to use the npm package

**Comment by silverwind** [comment]
`node_modules/material-icon-theme/dist/material-icons.json` won't be present I guess, so it needs to be generated otherwise. I do hope we get https://github.com/PKief/vscode-material-icon-theme/issues/1945 fixed soon, as it seems to be the only missing puzzle piece.

**Comment by lunny** [comment]
1 It's difficult to differ folder icon with others
2 README.md has a diffierent icon as other markdown files.
3 I personally like the previous icons if it could be the default it's better but this is a blocker. Or it can be a user level options in future.

**Comment by yardenshoham** [comment]
If I change the default to the previous icons then will that be fine? People will have to opt-in so points 1 and 2 don't hold as much weight

I wanted to make this a user setting but I don't want to bloat this PR too much

**Comment by yardenshoham** [comment]
I changed the default back to `basic`

**Comment by wxiaoguang** [comment]
I haven't looked into details, however ....

> I changed the default back to `basic`

If it's default to `basic`, then 99.99% users won't know/use/benefit from it. Few people would read the config manual line by line and change this option.

**Comment by yardenshoham** [comment]
Honestly, I prefer `file-specific` as the default but I also know I need 2 `lgtm`s...

**Comment by KN4CK3R** [comment]
I'm not a fan. In a windows explorer list the icon gives you a hint which app will be started to open/run the file. Gitea opens every file as text file (except the few files we render: images, markdown, csv, ...). So a different icon adds no real value in my opinion.

**Comment by wxiaoguang** [comment]
Actually, I also dislike making the UI too colorful. UI color design should be restraint instead of mixing too many colors together.  Too many colors reduces the clarity of the UI, and increasing reading difficulty.

**Comment by yardenshoham** [comment]
I see. https://github.com/go-gitea/gitea/issues/11149 was open with 10 positive reactions so I thought it'd be something people want. Should I close this PR and the feature request issue?

**Review by lunny — DISMISSED** [review]

**Comment by yardenshoham** [comment]
Now that's a twist

**Review by delvh — APPROVED** [review]
Welcome to what I predicted above :grin:.
 
I agree with @wxiaoguang and @yardenshoham that an opt-out mechanism is better here than opt-in:
Otherwise, users simply won't know that it exists.
It had a reason that I suggested `file-specific` as a default.
If users don't like it, they can turn it off.

However, I'm approving now as the feature works in both cases and I don't want to hold this feature back.
I'd still like it if the default value was switched again.

File: modules/icon/icon.go
Comment (wxiaoguang) [inline modules/icon/icon.go:19]: This type is exported, but I guess it only used in this package internally?

File: modules/icon/icon.go
Comment (wxiaoguang) [inline modules/icon/icon.go:52]: It's better to make logs have some context texts or prefixes. Otherwise it's different to know where the log comes from if a user sees the "err.Error()"  string.

File: modules/icon/icon.go
Comment (wxiaoguang) [inline modules/icon/icon.go:73]: Here and below, it doesn't seem a Golang style.

```go
if name, ok := theMap[key]; ok {
    return name
}
```

**Comment by yardenshoham** [comment]
I agree with you @delvh but _we don't have the votes_. Unless someone else wants to LGTM with the `file-specific` as default?

File: build/generate-svg.js
Comment (wxiaoguang) [inline build/generate-svg.js:14]: Should `material-icons.json` also be processed to remove these entries?

And does the `iconDefinitions` key could be removed?

File: build/generate-svg.js
Comment (wxiaoguang) [inline build/generate-svg.js:25]: Could  there be a comment about how much size does this icon package increase?

**Comment by wxiaoguang** [comment]
> I see. #11149 was open with 10 positive reactions so I thought it'd be something people want. Should I close this PR and the feature request issue?

I guess people would prefer customizable icons, instead of the builtin.

If two user say "I like colorful icons", they may like different color styles, or different coloring rules. If the colors/icons are not customizable, in the future, there would be issues like "the icon is wrong" "I need that icon", then we do not have a ability to handle these requirements.

So, that's what I mean about "it should be re-evaluated about ..... or other solutions (theme plugin? custom template?)"

The most flexible solution is let users make the "customized icon package" in their "custom" directory, and they make their own styles. There are so many different file types, IMO we can't support them by this hard-coded builtin icon package.

**Comment by KN4CK3R** [comment]
How about "if you want this, install https://github.com/Claudiohbsantos/github-material-icons-extension"?

**Comment by wxiaoguang** [comment]
> to render server-side. Avoid flashes of unstyled content.

Let users make the "customized icon package" in their "custom" directory, then these icons are still loaded from backend.

**Comment by yardenshoham** [comment]
We'd have to build a whole system to map filenames to icons. This is better, simpler, and we don't force it on anyone (it's a setting)

**Comment by wxiaoguang** [comment]
> We'd have to build a whole system to map filenames to icons. This is better, simpler, and we don't force it on anyone (it's a setting)

But how to handle these cases: 

* Some files are recognized wrong and use non-ideal icons.
* Some users want to use other style icons (eg: do not use icons for some directories).
* Some private instances may have their own private file types and need private icons.

**Comment by KN4CK3R** [comment]
I like the "icon pack" idea. Add files with the name `cpp.svg` or something in a directory and afterwards `.cpp` files get that icon. By default that folder is empty which results in the default icon.

**Comment by delvh** [comment]
Hmm, yes.
As much as I hate to retract my approval, the icon pack approach is probably a much better implementation long-term.
The good thing is: once it is done, there are no maintenance costs for us as the admins are responsible themselves to keep it up to date.

**Comment by yardenshoham** [comment]
By that logic, we can drop all SVGs...

**Comment by delvh** [comment]
Well… Yes.
As I've already written on Discord, I'm also fine with marking the functionality as deprecating/experimental and subject to change, and then refactoring the mechanism afterward.
Given those circumstances, I can re-approve.

**Comment by yardenshoham** [comment]
Experimental is fine I guess

**Comment by lunny** [comment]
> > We'd have to build a whole system to map filenames to icons. This is better, simpler, and we don't force it on anyone (it's a setting)
> 
> But how to handle these cases:
> 
>     * Some files are recognized wrong and use non-ideal icons.
> 
>     * Some users want to use other style icons (eg: do not use icons for some directories).
> 
>     * Some private instances may have their own private file types and need private icons.
> 
> 
> My proposal could include the "rules" json file, keep the icon-detection algorithm here
> 
> To be "simple", Gitea could provide an icon package officially, users just need to download it.

Good idea. Could that become a part of theme?

**Comment by yardenshoham** [comment]
I understand this approach will not be accepted

**Comment by wxiaoguang** [comment]
@yardenshoham  if you don't mind, I would try to implement "icon pack" approach, and re-use your code (co-author), does it look good to you?

**Comment by yardenshoham** [comment]
Yes, go ahead

**Comment by silverwind** [comment]
I wouldn't mind if we ship a default, unopinionated pack of icons. Requiring extra configuration and especially defining a plugin interface for icons seems too much work.

**Comment by yardenshoham** [comment]
That is to say, the approach in this PR was good enough?

**Comment by lunny** [comment]
> I wouldn't mind if we ship a default, unopinionated pack of icons that can be turned off via an option. Requiring extra configuration and especially defining a plugin interface for icons seems too much work and it would probably not see much adoption by the community I guess.

But if users could put their icons on `custom/`, it could be part of the theme which is attractive.

**Comment by yardenshoham** [comment]
Still, having the material icons by default would look good to most people that need no customization

**Comment by silverwind** [comment]
I can accept this PR. I do not see icons as part of the theme, but if we can make them overridable via custom folder, it's a nice bonus.

I wouldn't expect any community themer to keep up with this task. Assuming we are going to frequently update the icons, it will be a herculean task to keep a custom icon set in sync with our base, which if it is the material one, seems to add new icons every week.

**Comment by yardenshoham** [comment]
I'm not gonna reopen because I think the majority will block

**Comment by silverwind** [comment]
> I like the "icon pack" idea. Add files with the name `cpp.svg` or something in a directory and afterwards `.cpp` files get that icon. By default that folder is empty which results in the default icon. Then everyone can add the desired icons or keep it empty. This wouldn't bloat compiled Gitea too.

It won't be so simple in many other cases where regex match is needed. That is one of the reasons why I don't see this as suitable for themeing.

I don't see it as bloating, I see it as a nice optional enhancement to the UI. What is the total file size of the icons?

**Comment by wxiaoguang** [comment]
I am trying to implement a "custom" approach based on yardenshoham's solution.

Let's see if my proposal could work. If I my proposal has problems, let's continue here.

**Comment by yardenshoham** [comment]
> What is the total file size of the icons?
```
2.8M    node_modules/material-icon-theme/icons
160K    assets/material-icons.json
```

**Comment by delvh** [comment]
> It won't be so simple in many other cases where regex match is needed

That hasn't even been implemented with the approach proposed in this PR, so I don't see a reason to support it at all.
How often does it happen that you actually need a regex to describe which files it applies to?
The only one that comes to mind is `ya?ml`.
And that simply needs the same icon for the two different values.

**Comment by yardenshoham** [comment]
I'm sure that the material theme is good enough for 80%+ of users and if not, it can be disabled...

**Comment by silverwind** [comment]
I hadn't checked in detail how the matching worked (because the diff view is so slow), but it didn't seem like a simple "endsWith" check at first glance, but maybe it actually is.

**Pull request closed without merging**
